src/Features/Core/Portable/ConvertAnonymousTypeToClass/AbstractConvertAnonymousTypeToClassCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,422 @@
+        {
+            var document = context.Document;
+            var cancellationToken = context.CancellationToken;
+
+            var (anonymousObject, anonymousType) = await TryGetAnonymousObjectAsync(
+                document, context.Span, cancellationToken).ConfigureAwait(false);
+
+            if (anonymousObject == null || anonymousType == null)
+            {
+                return;
+            }
+
+            // Check if the anonymous type actually references another anonymous type inside of it.
+            // If it does, we can't convert this.  There is no way to describe this anonymous type
+            // in the concrete type we create.
+            var containsAnonymousType = anonymousType.GetMembers()
+                                                     .OfType<IPropertySymbol>()
+                                                     .Any(p => p.Type.ContainsAnonymousType());
+            if (containsAnonymousType)
+            {
+                return;
+            }
+
+            context.RegisterRefactoring(new MyCodeAction(
+                c => ConvertToClassAsync(document, context.Span, c)));
+        }
+
+        private async Task<(TAnonymousObjectCreationExpressionSyntax, INamedTypeSymbol)> TryGetAnonymousObjectAsync(
+            Document document, TextSpan span, CancellationToken cancellationToken)
+        {
+            var position = span.Start;
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var token = root.FindToken(position);
+
+            // Span actually has to be within the token (i.e. not in trivia around it).
+            if (!token.Span.IntersectsWith(position))
+            {
+                return default;
+            }
+
+            if (!span.IsEmpty && span != token.Span)
+            {
+                // if there is a selection, it has to be of the whole token.
+                return default;
+            }
+
+            var anonymousObject = token.Parent as TAnonymousObjectCreationExpressionSyntax;
+            if (anonymousObject == null)
+            {
+                return default;
+            }
+
+            // The position/selection must be of the 'new' token of the anonymous object.
+            if (anonymousObject.GetFirstToken() != token)
+            {
+                return default;
+            }
+
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var anonymousType = semanticModel.GetTypeInfo(anonymousObject, cancellationToken).Type as INamedTypeSymbol;
+
+            return (anonymousObject, anonymousType);
+        }
+
+        private async Task<Document> ConvertToClassAsync(
+            Document document, TextSpan span, CancellationToken cancellationToken)
+        {
+            var (anonymousObject, anonymousType) = await TryGetAnonymousObjectAsync(
+                document, span, cancellationToken).ConfigureAwait(false);
```

Comment:
Should not we pass (anonymousObject, anonymousType) from to ConvertToClassAsync as a parameter? It seems we execute TryGetAnonymousObjectAsync twice.

Reply:
No.  One computation is in deciding to show the lightbulb.  The other is in actually generating the type when the lightbulb is invoked.  We don't want lightbulbs holding onto expensive resources that might keep a large swath of the symbol graph alive.  So this is our fairly standard pattern of going and re-fetching the data once invoked.